operator/internal/controller/pulumi/git.go

```diff
@@ -18,318 +18,276 @@ import (
+		return nil, fmt.Errorf(`missing "commit" or "branch"`)
+	}
+	if gs.Commit != "" && gs.Branch != "" {
+		return nil, fmt.Errorf(`only one of "commit" or "branch" should be specified`)
+	}
+
+	ref := gs.Commit
+	if gs.Branch != "" {
+		ref = gs.Branch
+	}
+	url := gs.ProjectRepo
+
+	return newGitSource(url, ref, auth)
+}
+
+func newGitSource(rawURL string, ref string, auth *auto.GitAuth) (Source, error) {
+	url, _, err := gitutil.ParseGitRepoURL(rawURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing %q: %w", rawURL, err)
+	}
 
+	fs := memory.NewStorage()
+	remote := git.NewRemote(fs, &config.RemoteConfig{
+		Name: "origin",
+		URLs: []string{url},
+	})
+
+	return &gitSource{
+		fs:     fs,
+		ref:    ref,
+		remote: remote,
+		auth:   auth,
+	}, nil
 }
 
-var transportMutex sync.Mutex
+type gitSource struct {
+	fs     *memory.Storage
+	ref    string
+	remote *git.Remote
+	auth   *auto.GitAuth
+}
 
-func setupGitRepo(ctx context.Context, workDir string, repoArgs *auto.GitRepo) (string, error) {
-	cloneOptions := &git.CloneOptions{
-		RemoteName: "origin", // be explicit so we can require it in remote refs
-		URL:        repoArgs.URL,
+func (gs gitSource) CurrentCommit(ctx context.Context) (string, error) {
+	// If our ref is already a commit then use it directly.
+	if plumbing.IsHash(gs.ref) {
+		return gs.ref, nil
 	}
 
-	if repoArgs.Shallow {
-		cloneOptions.Depth = 1
-		cloneOptions.SingleBranch = true
+	// Otherwise fetch the most recent commit for the ref (branch) we care
+	// about.
+	auth, err := gs.authMethod()
+	if err != nil {
+		return "", fmt.Errorf("getting auth methodL: %w", err)
 	}
 
-	if repoArgs.Auth != nil {
-		authDetails := repoArgs.Auth
-		// Each of the authentication options are mutually exclusive so let's check that only 1 is specified
-		if authDetails.SSHPrivateKeyPath != "" && authDetails.Username != "" ||
-			authDetails.PersonalAccessToken != "" && authDetails.Username != "" ||
-			authDetails.PersonalAccessToken != "" && authDetails.SSHPrivateKeyPath != "" ||
-			authDetails.Username != "" && authDetails.SSHPrivateKey != "" {
-			return "", errors.New("please specify one authentication option of `Personal Access Token`, " +
-				"`Username\\Password`, `SSH Private Key Path` or `SSH Private Key`")
+	refs, err := gs.remote.ListContext(ctx, &git.ListOptions{
+		Auth: auth,
+	})
+	if err != nil {
+		return "", fmt.Errorf("listing: %w", err)
+	}
+	for _, r := range refs {
+		if r.Name().String() == gs.ref || r.Name().Short() == gs.ref {
+			return r.Hash().String(), nil
 		}
+	}
 
-		// Firstly we will try to check that an SSH Private Key Path has been specified
-		if authDetails.SSHPrivateKeyPath != "" {
-			publicKeys, err := ssh.NewPublicKeysFromFile("git", repoArgs.Auth.SSHPrivateKeyPath, repoArgs.Auth.Password)
-			if err != nil {
-				return "", fmt.Errorf("unable to use SSH Private Key Path: %w", err)
-			}
+	return "", fmt.Errorf("no commits found for ref %q", gs.ref)
```

Comment:
When this (or similar spec problem) happens, does the status transition to Stalled as it should?

Reply:
Yes, with SourceUnavailable.